battle.go

```diff
@@ -1,115 +1,212 @@
 package pokemonbattlelib
 
 import (
-	"log"
+	"fmt"
 )
 
 type Battle struct {
-	Weather  int  // one of the 6 in-battle weather conditions
-	ShiftSet bool // shift or set battle style for NPC trainer battles
-	State    BattleState
-	Players  []*Agent
+	Weather      int  // one of the 6 in-battle weather conditions
+	ShiftSet     bool // shift or set battle style for NPC trainer battles
+	State        BattleState
+	Parties      []*Party
+	Agents       []*Agent
+	agentParties map[int]int // Maps agent indexes to party indexes
+	teams        [][]int     // An array of teams, which are arrays of Agents. Used to derive allies and opponents
 }
 
-type BattleState int
+type Party struct {
+	Pokemon []*Pokemon
+	active  []int
+}
 
-type Player int
+func (p *Party) AddPokemon(pkmn ...*Pokemon) {
+	p.Pokemon = append(p.Pokemon, pkmn...)
+}
+
+func (p *Party) SetActive(idx int) {
+	if len(p.active) == 0 {
+		p.active = append(p.active, idx)
+	} else {
+		p.active[0] = idx
+	}
+}
+
+func (p *Party) GetActive() []int {
+	return p.active
+}
+
+// Gives an Agent control over a Party.
+func (b *Battle) LinkAgentParty(agentIdx, partyIdx int) error {
+	if b.agentParties == nil {
+		b.agentParties = make(map[int]int)
+	}
+	// TODO: check if agentIdx, partyIdx are valid indexes
+	// TODO: check if agent is already linked
+	// TODO: check if party is already linked
+	b.agentParties[agentIdx] = partyIdx
+	return nil
+}
+
+// Set which agents have which allies. Not to be confused with `Party`.
+func (b *Battle) SetTeams(t [][]int) error {
+	// TODO: validate
+	b.teams = t
+	return nil
+}
+
+type BattleState int
 
-const (
-	PLAYER1 Player = iota
-	PLAYER2
-)
 const (
 	BEFORE_START BattleState = iota
-	BATTLE_START
-	BEFORE_PLAYER1_TURN
-	AFTER_PLAYER1_TURN
-	BEFORE_PLAYER2_TURN
-	AFTER_PLAYER2_TURN
+	BATTLE_IN_PROGRESS
 	BATTLE_END
 )
```

Comment:
In a round of a Pokemon battle, turns for each pokemon are queued up, and then the game decides what order they execute in. There is no need for the agent to know when turns are happening, because the agent is only responsible for making decisions based on input.